app.go

```diff
@@ -24,142 +24,238 @@ import (
 	"context"
 	"os"
 	"os/signal"
-	"reflect"
 	"syscall"
 	"time"
 
-	"github.com/pkg/errors"
 	"go.uber.org/dig"
 	"go.uber.org/fx/internal/fxlog"
 	"go.uber.org/fx/internal/fxreflect"
 	"go.uber.org/multierr"
 )
 
-// App models a modular application
+// DefaultTimeout is the start and stop timeout used by Run.
+const DefaultTimeout = 15 * time.Second
+
+// Timeout is a convenience function to construct a context with a timeout.
+// It's only intended to reduce noise in the main function; since it doesn't
+// expose context.CancelFunc, it may leak resources.
+func Timeout(d time.Duration) context.Context {
+	ctx, cancel := context.WithTimeout(context.Background(), d)
+	// Assign to the blank identifier on a separate line to convince the linter
+	// that we really don't want the cancel function.
+	_ = cancel
+	return ctx
+}
+
+// An Option configures an App.
+type Option interface {
+	apply(*App)
+}
+
+type optionFunc func(*App)
+
+func (f optionFunc) apply(app *App) { f(app) }
+
+// Provide registers constructors with the application's dependency injection
+// container. Constructors provide one or more types, can depend on other
+// types available in the container, and may optionally return an error. For
+// example:
+//
+//  // Provides type *C, depends on *A and *B.
+//  func(*A, *B) *C
+//
+//  // Provides type *C, depends on *A and *B, and indicates failure by
+//  // returning an error.
+//  func(*A, *B) (*C, error)
+//
+//  // Provides type *B and *C, depends on *A, and can fail.
+//  func(*A) (*B, *C, error)
+//
+// The order in which constructors are provided doesn't matter. Constructors
+// are called lazily and their results are cached for reuse.
+//
+// Taken together, these properties make it perfectly reasonable to Provide a
+// large number of standard constructors even if only a fraction of them are
+// used.
+//
+// See the documentation for go.uber.org/dig for further details.
+func Provide(constructors ...interface{}) Option {
+	return optionFunc(func(app *App) {
+		for _, c := range constructors {
+			app.provide(c)
+		}
+	})
+}
+
+// Invoke registers functions that are executed eagerly on application start.
+// Arguments for these functions are provided from the application's
+// dependency injection container, so the arguments form the leaves of the
+// dependency resolution graph. That is, only constructors whose returned
```

Comment:
nit: Complexity consistency. The docs for Provide are very approachable but the docs for Invoke are throwing graph resolution at the user.